TensorKitSectors/src/product.jl

```diff
@@ -29,25 +30,24 @@ function Base.size(::SectorValues{ProductSector{T}}) where {T<:SectorTuple}
 end
 Base.length(P::SectorValues{<:ProductSector}) = *(size(P)...)
 
-function Base.iterate(::SectorValues{ProductSector{T}}, args...) where {T<:SectorTuple}
-    next = iterate(product(values.(_sectors(T))...), args...)
-    next === nothing && return nothing
-    val, state = next
-    return ProductSector{T}(val), state
+function _length(iter::SectorValues{I}) where {I<:Sector}
+    return Base.IteratorSize(iter) === Base.IsInfinite() ? typemax(Int) : length(iter)
+end
+function _size(::SectorValues{ProductSector{T}}) where {T<:SectorTuple}
+    return map(s -> _length(values(s)), _sectors(T))
 end
 function Base.getindex(P::SectorValues{ProductSector{T}}, i::Int) where {T<:SectorTuple}
-    Base.IteratorSize(P) isa IsInfinite &&
-        throw(ArgumentError("cannot index into infinite product sector"))
-    return ProductSector{T}(getindex.(values.(_sectors(T)),
-                                      Tuple(CartesianIndices(size(P))[i])))
+    I = manhattan_to_multidimensional_index(i, _size(P))
+    return ProductSector{T}(getindex.(values.(_sectors(T)), I))
 end
 function findindex(P::SectorValues{ProductSector{T}},
-                   c::ProductSector{T}) where
-         {T<:SectorTuple}
-    Base.IteratorSize(P) isa IsInfinite &&
-        throw(ArgumentError("cannot index into infinite product sector"))
-    return LinearIndices(size(P))[CartesianIndex(findindex.(values.(_sectors(T)),
-                                                            c.sectors))]
+                   c::ProductSector{T}) where {T<:SectorTuple}
+    return to_manhattan_index(findindex.(values.(_sectors(T)), c.sectors), _size(P))
+end
+
+function Base.iterate(P::SectorValues{ProductSector{T}}, i=1) where {T<:SectorTuple}
+    Base.IteratorSize(P) != Base.IsInfinite() && i > length(P) && return nothing
+    return getindex(P, i), i + 1
```

Comment:
I guess here is the biggest possible speed trap, it feels like it could be possible to iterate through these without having to map to linear indices, similar to how cartesianindices can be iterated without the integer divisions.

Reply:
Given that iteration over all possible sector values is not really used in the code, except for making tests, I don't care too much about the speed of this :-)